lib/fuse_uring.c

```diff
@@ -0,0 +1,571 @@
+};
+
+static size_t
+fuse_ring_queue_size(const size_t q_depth)
+{
+	const size_t req_size = sizeof(struct fuse_ring_ent) * q_depth;
+
+	return sizeof(struct fuse_ring_queue) + req_size;
+}
+
+static struct fuse_ring_queue *
+fuse_uring_get_queue(struct fuse_ring_pool *fuse_ring, int qid)
+{
+	char *ptr =
+		((char *)fuse_ring->queues) + (qid * fuse_ring->queue_mem_size);
+
+	return (struct fuse_ring_queue *)ptr;
+}
+
+/**
+ * return a pointer to the 80B area
+ */
+static void *fuse_uring_get_sqe_cmd(struct io_uring_sqe *sqe)
+{
+	return (void *)&sqe->cmd[0];
+}
+
+static void fuse_uring_sqe_set_req_data(struct fuse_uring_cmd_req *req,
+					const unsigned int qid,
+					const unsigned int commit_id)
+{
+	req->qid = qid;
+	req->commit_id = commit_id;
+	req->flags = 0;
+}
+
+static void
+fuse_uring_sqe_prepare(struct io_uring_sqe *sqe, struct fuse_ring_ent *req,
+		       __u32 cmd_op)
+{
+	/* These fields should be written once, never change */
+	sqe->opcode = IORING_OP_URING_CMD;
+
+	/*
+	 * IOSQE_FIXED_FILE: fd is the index to the fd *array*
+	 * given to io_uring_register_files()
+	 */
+	sqe->flags = IOSQE_FIXED_FILE;
+	sqe->fd = 0;
+
+	sqe->rw_flags = 0;
+	sqe->ioprio = 0;
+	sqe->off = 0;
+
+	io_uring_sqe_set_data(sqe, req);
+
+	sqe->cmd_op = cmd_op;
+	sqe->__pad1 = 0;
+}
+
+static int fuse_queue_setup_io_uring(struct io_uring *ring, size_t qid,
+				     size_t depth, int fd, int evfd)
+{
+	int rc;
+	struct io_uring_params params = {0};
+	int files[2] = { fd, evfd };
+
+	depth += 1; /* for the eventfd poll SQE */
+
+	params.flags = IORING_SETUP_SQE128;
+
+	/* Avoid cq overflow */
+	params.flags |= IORING_SETUP_CQSIZE;
+	params.cq_entries = depth * 2;
+
+	/* These flags should help to increase performance, but actually
+	 * make it a bit slower - reason should get investigated.
+	 */
+	if (0) {
+		/* Has the main slow down effect */
+		params.flags |= IORING_SETUP_SINGLE_ISSUER;
+
+		// params.flags |= IORING_SETUP_DEFER_TASKRUN;
+		params.flags |= IORING_SETUP_TASKRUN_FLAG;
+
+		/* Second main effect to make it slower */
+		params.flags |= IORING_SETUP_COOP_TASKRUN;
+	}
+
+	rc = io_uring_queue_init_params(depth, ring, &params);
+	if (rc != 0) {
+		fuse_log(FUSE_LOG_ERR, "Failed to setup qid %zu: %d (%s)\n",
+			 qid, rc, strerror(-rc));
+		return rc;
+	}
+
+	rc = io_uring_register_files(ring, files, 1);
+	if (rc != 0) {
+		rc = -errno;
+		fuse_log(FUSE_LOG_ERR,
+			 "Failed to register files for ring idx %zu: %s",
+			 qid, strerror(errno));
+		return rc;
+	}
+
+	return 0;
+}
+
+static void free_payload(struct fuse_ring_ent *ent)
+{
+	numa_free(ent->op_payload, ent->req_payload_sz);
+}
+
+static void fuse_session_destruct_uring(struct fuse_ring_pool *fuse_ring)
+{
+	for (size_t qid = 0; qid < fuse_ring->nr_queues; qid++) {
+		struct fuse_ring_queue *queue =
+			fuse_uring_get_queue(fuse_ring, qid);
+
+		if (queue->tid != 0)
+			pthread_join(queue->tid, NULL);
+
+		if (queue->ring.ring_fd != -1)
+			close(queue->ring.ring_fd);
```

Comment:
Better to destruct the uring with io_uring_queue_exit() helper?

Reply:
Ah nice, I didn't know this function. A bit a pity that there are no checks if the ring is initialized at all

```__cold void io_uring_queue_exit(struct io_uring *ring)
{
    struct io_uring_sq *sq = &ring->sq;
    struct io_uring_cq *cq = &ring->cq;
    size_t sqe_size;
 
    if (!sq->ring_sz && !(ring->int_flags & INT_FLAG_APP_MEM)) {
        sqe_size = sizeof(struct io_uring_sqe);
        if (ring->flags & IORING_SETUP_SQE128)
            sqe_size += 64;
```

I.e. we need to keep the check for ring_fd == -1.

Reply:
"there are no checks if the ring is initialized at all"

Yeah,  that's true.